From c099458e3594fe47c510d7c2b1aa9be2ea19c67f Mon Sep 17 00:00:00 2001
From: Sami Shalayel <sami.shalayel@qt.io>
Date: Fri, 10 Oct 2025 11:17:21 +0200
Subject: [PATCH 155/239] Qt.rgba: use qBound to avoid asserts

Use qBound that knows how to handle Nan to clip values into the 0, 1.0
range. The other methods (hsva, hsla) also use qBound to bound their
values.

This avoids an assert later when trying to round the NaN during the
color conversion. The assert seems to have been added by
df97b6b2de6282bd6422f1e531a42475dadc980d.

Task-number: QTBUG-140899
Change-Id: I4bb34b92ceac9d3ad667ae9b2714772bc800a600
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
Reviewed-by: Ulf Hermann <ulf.hermann@qt.io>
Reviewed-by: Semih Yavuz <semih.yavuz@qt.io>
(cherry picked from commit 3f9522b233798e9f79368aeab10639f35dcb6d3a)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/qml/qml/qqmlbuiltinfunctions.cpp | 12 ++++--------
 tests/auto/qml/qqmlqt/data/hsla.qml  |  1 +
 tests/auto/qml/qqmlqt/data/hsva.qml  |  1 +
 tests/auto/qml/qqmlqt/data/rgba.qml  |  1 +
 tests/auto/qml/qqmlqt/tst_qqmlqt.cpp |  3 +++
 5 files changed, 10 insertions(+), 8 deletions(-)

diff --git a/src/qml/qml/qqmlbuiltinfunctions.cpp b/src/qml/qml/qqmlbuiltinfunctions.cpp
index 12f3f27845..70926903ba 100644
--- a/src/qml/qml/qqmlbuiltinfunctions.cpp
+++ b/src/qml/qml/qqmlbuiltinfunctions.cpp
@@ -339,14 +339,10 @@ QVariant QtObject::color(const QString &name) const
 */
 QVariant QtObject::rgba(double r, double g, double b, double a) const
 {
-    if (r < 0.0) r=0.0;
-    if (r > 1.0) r=1.0;
-    if (g < 0.0) g=0.0;
-    if (g > 1.0) g=1.0;
-    if (b < 0.0) b=0.0;
-    if (b > 1.0) b=1.0;
-    if (a < 0.0) a=0.0;
-    if (a > 1.0) a=1.0;
+    r = qBound(0.0, r, 1.0);
+    g = qBound(0.0, g, 1.0);
+    b = qBound(0.0, b, 1.0);
+    a = qBound(0.0, a, 1.0);
 
     return QQml_colorProvider()->fromRgbF(r, g, b, a);
 }
diff --git a/tests/auto/qml/qqmlqt/data/hsla.qml b/tests/auto/qml/qqmlqt/data/hsla.qml
index 70dc87d5e0..4403d2f701 100644
--- a/tests/auto/qml/qqmlqt/data/hsla.qml
+++ b/tests/auto/qml/qqmlqt/data/hsla.qml
@@ -8,5 +8,6 @@ QtObject {
     property color test5: Qt.hsla(1.2, 1.3, 1.4, 1.5);
     property color test6: Qt.hsla(-0.1, -0.2, -0.3, -0.4);
     property color test7: Qt.hsla(-1, 0, 0.5, 1);
+    property color test8: Qt.hsla(undefined, undefined, undefined, undefined); // should not crash
 }
 
diff --git a/tests/auto/qml/qqmlqt/data/hsva.qml b/tests/auto/qml/qqmlqt/data/hsva.qml
index 8c9a2a56ff..2ffd0a064e 100644
--- a/tests/auto/qml/qqmlqt/data/hsva.qml
+++ b/tests/auto/qml/qqmlqt/data/hsva.qml
@@ -8,5 +8,6 @@ QtObject {
     property color test5: Qt.hsva(1.2, 1.3, 1.4, 1.5);
     property color test6: Qt.hsva(-0.1, -0.2, -0.3, -0.4);
     property color test7: Qt.hsva(-1, 0, 0.5, 1);
+    property color test8: Qt.hsva(undefined, undefined, undefined, undefined); // should not crash
 }
 
diff --git a/tests/auto/qml/qqmlqt/data/rgba.qml b/tests/auto/qml/qqmlqt/data/rgba.qml
index 3b010f68cb..3c3d022e4d 100644
--- a/tests/auto/qml/qqmlqt/data/rgba.qml
+++ b/tests/auto/qml/qqmlqt/data/rgba.qml
@@ -7,4 +7,5 @@ QtObject {
     property color test4: Qt.rgba(1, 1, 1, 1, 1);
     property color test5: Qt.rgba(1.2, 1.3, 1.4, 1.5);
     property color test6: Qt.rgba(-0.1, -0.2, -0.3, -0.4);
+    property color test7: Qt.rgba(undefined, undefined, undefined, undefined); // should not crash
 }
diff --git a/tests/auto/qml/qqmlqt/tst_qqmlqt.cpp b/tests/auto/qml/qqmlqt/tst_qqmlqt.cpp
index bb7ce5fe65..4446768f06 100644
--- a/tests/auto/qml/qqmlqt/tst_qqmlqt.cpp
+++ b/tests/auto/qml/qqmlqt/tst_qqmlqt.cpp
@@ -178,6 +178,7 @@ void tst_qqmlqt::rgba()
     QCOMPARE(qvariant_cast<QColor>(object->property("test4")), QColor());
     QCOMPARE(qvariant_cast<QColor>(object->property("test5")), QColor::fromRgbF(1, 1, 1, 1));
     QCOMPARE(qvariant_cast<QColor>(object->property("test6")), QColor::fromRgbF(0, 0, 0, 0));
+    QCOMPARE(qvariant_cast<QColor>(object->property("test7")), QColor::fromRgbF(0, 0, 0, 0));
 }
 
 void tst_qqmlqt::hsla()
@@ -201,6 +202,7 @@ void tst_qqmlqt::hsla()
     QColor test7 = qvariant_cast<QColor>(object->property("test7"));
     QCOMPARE(test7, QColor::fromHslF(-1, 0, 0.5, 1));
     QCOMPARE(test7.hslHue(), -1.0f);
+    QCOMPARE(qvariant_cast<QColor>(object->property("test8")), QColor::fromHslF(0, 0, 0, 0));
 }
 
 void tst_qqmlqt::hsva()
@@ -224,6 +226,7 @@ void tst_qqmlqt::hsva()
     QColor test7 = qvariant_cast<QColor>(object->property("test7"));
     QCOMPARE(test7, QColor::fromHsvF(-1, 0, 0.5, 1));
     QCOMPARE(test7.hsvHue(), -1.0f);
+    QCOMPARE(qvariant_cast<QColor>(object->property("test8")), QColor::fromHsvF(0, 0, 0, 0));
 }
 
 void tst_qqmlqt::colorEqual()
-- 
2.51.2

